[0.20] Backports

fanquake opened the pull request:
Backports the following PRs to the 0.20 branch:
* https://github.com/bitcoin/bitcoin/pull/19777 - docs: Correct description for getblockstats's txs field
* https://github.com/bitcoin/bitcoin/pull/19836 - rpc: Properly deserialize txs with witness before signing
* https://github.com/bitcoin/bitcoin/pull/20080 - Strip any trailing `/` in -datadir and -blocksdir paths
* https://github.com/bitcoin/bitcoin/pull/20082 - [bugfix] random: fixes read buffer to use min rather than max
* https://github.com/bitcoin/bitcoin/pull/20141 - Avoid the use of abs64 in timedata
* https://github.com/bitcoin/bitcoin/pull/20146 - net: Send post-verack handshake messages at most once
* https://github.com/bitcoin/bitcoin/pull/20195 - build: fix mutex detection when building bdb on macOS
* https://github.com/bitcoin/bitcoin/pull/20298 - macOS deploy: use the new plistlib API

Will add additional commits as they become available.

shesek committed:
docs: Correct description for getblockstats's txs field

It does count the coinbase transaction.

Refs #19766

Github-Pull: #19777
Rebased-From: 4148f55dd016f940df50a44cf03d117cdb1dd929

Commit:
rpc: Properly deserialize txs with witness before signing

Github-Pull: #19836
Rebased-From: cccc7525697e7b8d99b545e34f0f504c78ffdb94

EthanHeilman committed:
random: fixes read buffer resizing in RandAddSeedPerfmon

+ Replaces std::max with std::min to resize buffer in RandAddSeedPerfmon
+ Documents behavior of RandAddSeedPerfmon

Github-Pull: #20082
Rebased-From: bd5215103eb3985c1622eddea45a040e6173829c

sipa committed:
Avoid the use of abs64 in timedata

Github-Pull: #20141
Rebased-From: d1292f25f272401da0c58580521c74b1fa03a9ad

Commit:
rpc: Adjust witness-tx deserialize error message

Github-Pull: #19836
Rebased-From: 33330778230961cfbf2a24de36b5877e395cc596

Commit:
net: Send post-verack handshake messages at most once

Github-Pull: #20146
Rebased-From: fa1f6f237d02265af616129402fa2b8a3019dda5

hebasto committed:
util: Add StripRedundantLastElementsOfPath function

Co-authored-by: saibato <saibato.naga@pm.me>
Co-authored-by: MarcoFalke <falke.marco@gmail.com>

Github-Pull: 20080
Rebased-From: b19e88230f0e93e95e883e65376963cb9c36f606

hebasto committed:
doc: Update data directory path comments

Github-Pull: #20080
Rebased-From: ad5cef5dfdd5802fc187a52e74d940a52f420a51

fanquake committed:
build: fix mutex detection when building bdb on macOS

Starting with the Clang shipped with Xcode 12, Apple has enabled
-Werror=implicit-function-declaration by default. This causes bdbs mutex
detection to fail when building on macOS (not cross-compiling):

checking for mutexes... UNIX/fcntl
configure: WARNING: NO SHARED LATCH IMPLEMENTATION FOUND FOR THIS PLATFORM.
configure: error: Unable to find a mutex implementation

as previously emitted warnings are being turned into errors. i.e:

error: implicitly declaring library function 'exit' with type 'void (int) __attribute__((noreturn))' [-Werror,-Wimplicit-function-declaration]

Append -Wno-error=implicit-function-declaration to cflags so that
-Wimplicit-function-declaration returns to being a warning, and the
configure checks will succeed.

Fixes #19411.

Github-Pull: #20195
Rebased-From: d0a829e9632379e42f0be5c554e3b692f0d14a95

jnewbery commented:
utACK 1f67a30e8374951997af924293c60eff56ae39ed

Manually verified all the cherry picks were the same as master.

jonasschnelli committed:
macOS deploy: use the new plistlib API

See https://docs.python.org/3/library/plistlib.html.
The new API was added in 3.4 and old removed in 3.9.

Github-Pull: 20298
Rebased-From: 04a69c200e0d18ae63c7e47898f85d1b4cb5c23d

MarcoFalke commented:
review ACK 7566af419f3e7060152ae062f5a578f543a16a5a 🗡

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 7566af419f3e7060152ae062f5a578f543a16a5a 🗡
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjV0wv/YkXQDPhIjXO8ZKqVRZyyMi+RQtx4FiU5hL/PJKfZg2UEogGrORfOdxdc
FZw3zI9h580R6b3fqZPclPv3n+jpaBKfF98m5TKTfm0dsVX9XPmH2brB8eKA+yDo
bck5dyrdkootE52vzq5fQwcd3jb2SliIaYvLUBPvrlTe3oqWTZTdg+tjRNplAjUa
0hPps6Sn2m/GXypk4ZqcuclFiG+OiRak4mPGY7G++eCxh+eFYdvEpPi0z7z6RZK5
9i/Bw0eZACX6i5kDRsfBENdwRKZn6aW6SaSDdOOxMC+IM4cq2vEo0EXHXm5D1mEl
x7RyISFUyLz4fL7Adgme/+Ppwc6jgWu70JpcZa0BU2CLex3O7Ek9Wck/unJx6wBx
dLFckqS8rLI/SYfcswWOqLR9xmIQapa5lifuLn3JKKWcFYSO6H2mQEnlNJCRaVt4
99uXmAece9icmR90xLwKyeEGQ4gNyPCiKE4+mywvsYJchk4P38NIU+q0RljoO/6P
elRfqGXN
=SLep
-----END PGP SIGNATURE-----
```

python: error while loading shared libraries: libpython3.8.so.1.0: cannot open shared object file: No such file or directory

</details>
